tap_dynamodb/sync_strategies/log_based.py

```diff
@@ -112,22 +99,22 @@ def sync_shard(shard, seq_number_bookmarks, streams_client, stream_arn, projecti
                     raise RuntimeError('Projection failed to apply: {}'.format(projection))
 
         record_message = singer.RecordMessage(stream=table_name,
-                                            record=record_message,
-                                            version=stream_version)
+                                              record=record_message,
+                                              version=stream_version)
         singer.write_message(record_message)
 
         rows_synced += 1
 
-        seq_number_bookmarks[shard['ShardId']] = record['dynamodb']['SequenceNumber']
-        state = singer.write_bookmark(state, table_name, 'shard_seq_numbers', seq_number_bookmarks)
-
-    if rows_synced > 0:
-        singer.write_state(state)
+        # Every 100 rows update bookmarks and write the state
+        if rows_synced % 100 == 0:
+            seq_number_bookmarks[shard['ShardId']] = record['dynamodb']['SequenceNumber']
+            state = singer.write_bookmark(state, table_name, 'shard_seq_numbers', seq_number_bookmarks)
+            singer.write_state(state)
```

Comment:
>The bookmark should probably be written outside of the if, with state emission inside the if

I'm confused. First of all, why write the state outside the if it's not updated? And what do you mean emission inside the if? I agree with everything else you wrote above. Can you clarify the confusing point a bit though?

Reply:
I think @cosimon addressed your comment, but let me know if he didn't.

Reply:
Yup! Looks good. To elaborate, pulling out the bookmark advancement and state writing in this way makes sure that the state object always has the latest thing in it, so a State write can be placed anywhere. It may not matter right now, but if the end sequence number matters at some point in the future, it's a bit safer in my mind.